### PR TITLE
SAF authorization APIs

### PR DESCRIPTION
**Pull request by plavjanik**
Implements all the points of #3. 

This PR implements:
- [x] Improved authentication error handling and diagnostics
  - Wrong credentials are reported as 401 without details if the user ID is valid
  - Expired password is reported to the user as 401 with additional details
  - Internal authentication error (SAF error or misconfiguration) are reported as 500 and logged as errors to the server log
  - Added integration tests
- [x] Java API to do authorization checks
- [x] Authorization implementation using the Java SAF APIs (`DATASET` resource class is not in the scope of this PR - https://github.com/zowe/sample-spring-boot-api-service/issues/47)
- [x] Check that the application has required permissions for authentication at startup (`BPX.SERVER`)
- [x] Markdown documentation for authorization checks
- [x] Spring integration of the security checks for REST APIs

I recommend starting the review by reading the documentation - https://github.com/zowe/sample-spring-boot-api-service/pull/45/files#diff-ba22a8700b747216eee9fc3be67ece71

### CODE REVIEW

File: zowe-rest-api-sample-spring/zowe-api.json [hunk 1]
```diff
@@ -43,6 +43,23 @@
...
+                },
+                "config/ca32/keystore.p12": {
+                    "target": "etc/keystore.p12",
+                    "binary": true
+                },
+                "config/ca32/truststore.p12": {
+                    "target": "etc/truststore.p12",
+                    "binary": true
+                }
+            }
```
Comment (plavjanik) [inline zowe-rest-api-sample-spring/zowe-api.json]: This needs to be removed.

File: zowe-rest-api-sample-spring/config/zos/application.yml.hbs [hunk 2]
```diff
@@ -3,6 +3,10 @@
 
 spring.profiles.active: https,diag,zos
 
+logging:
+    level:
+        org.zowe.commons.zos.security: DEBUG
```
Comment (plavjanik) [inline zowe-rest-api-sample-spring/config/zos/application.yml.hbs]: Change to `INFO`

File: zowe-rest-api-commons-spring/src/main/resources/test-saf-invalid.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+safAccess:
```
Comment (plavjanik) [inline zowe-rest-api-commons-spring/src/main/resources/test-saf-invalid.yml]: Add comment to explain the file

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/spring/security/SafMethodSecurityExpressionRoot.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    public SafMethodSecurityExpressionRoot(Authentication authentication, PlatformSecurityService platformSecurityService, SafSecurityConfigurationProperties safSecurityConfigurationProperties) {
+        super(authentication);
+        this.platformSecurityService = platformSecurityService;
+        this.safSecurityConfigurationProperties = safSecurityConfigurationProperties;
+    }
+
+    public boolean hasSafResourceAccess(String resourceClass, String resourceName, String accessLevel) {
+        String userid = authentication.getName();
+        AccessLevel level = AccessLevel.valueOf(accessLevel);
+        log.debug("Evaluating access of user {} to resource {} in class {} level {}", userid, resourceName, level);
```
Comment (dkelosky) [inline zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/spring/security/SafMethodSecurityExpressionRoot.java]: i think this is missing a field substitution - class?

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java [hunk 5]
```diff
@@ -0,0 +1,1807 @@
...
+import java.util.Map;
+
+import lombok.Generated;
+
+/**
+ * Provides explanation for secondary error codes for as described at:
+ * https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.4.0/com.ibm.zos.v2r4.bpxa800/errnojrs.htm
+ */
+@Generated
+public enum PlatformErrno2 {
```
Comment (dkelosky) [inline zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java]: this seems useful to have these - i suppose in the future if IBM adds to their list - we'll just open PRs to update this as needed?

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/service/AccessControlService.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            } else {
+                message = "Platform access control failed: " + errno2.explanation;
+                if (!resourceHasToExist && (errno2 == PlatformErrno2.JRSAFResourceUndefined)) {
+                    return true;
+                } else if (errno2 == PlatformErrno2.JRNoResourceAccess) {
+                    return false;
+                } else if (errno2 == PlatformErrno2.JRSAFResourceUndefined) {
+                    return false;
+                }
+                log.error("Platform access control failed: {} {} {}", errno.name, errno2.name, errno2.explanation,
```
Comment (dkelosky) [inline zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/service/AccessControlService.java]: perhaps need another `{}` for substitution here?

File: zowe-rest-api-sample-spring/docs/zos-security.md [hunk 7]
```diff
@@ -199,3 +279,45 @@ The shared library `libsecur.so` is expected to be in zFS filesystem in the a di
...
+
+It provides following methods:
+
+- `boolean checkPermission(String userid, String resourceClass, String resourceName, AccessLevel accessLevel, boolean resourceHasToExist)`
+- `boolean checkPermission(String userid, String resourceClass, String resourceName, AccessLevel accessLevel)`
+- `boolean checkPermission(String resourceClass, String resourceName, AccessLevel accessLevel, boolean resourceHasToExist)`
+- `boolean checkPermission(String resourceClass, String resourceName, AccessLevel accessLevel)`
+
+All of them return `true` if user has access to the resource and `false` if the user has not access to the resource or the resource does not exist. They throw `AccessControlError` in other cases of failures (invalid user ID, invalid resource name, internal SAF error, misconfiguration of the service). The optional parameter `resourceHasToExist` is by default `true` and it causes that non-existing resource means that that no user has access to it. Some applications may want the opposite, protect access only when the security administator has created the resouces, then they can use `false` as the value of `resourceHasToExist` parameter.
```
Comment (dkelosky) [inline zowe-rest-api-sample-spring/docs/zos-security.md]: i think there is a typo here in wording `it causes that non-existing resource means that that no user has access to it`

File: zowe-rest-api-sample-spring/docs/zos-security.md [hunk 8]
```diff
@@ -199,3 +279,45 @@ The shared library `libsecur.so` is expected to be in zFS filesystem in the a di
...
+The SDK defined two new securiry expressions:
+
+- `boolean hasSafResourceAccess(String resourceClass, String resourceName, String accessLevel)` - return `true` when user has access to the resource
+- `boolean hasSafServiceResourceAccess(String resourceNameSuffix, String accessLevel)` - similar as the previous one but the resource class and resource name prefix is taken from the service configuration in `application.yml` under key `zowe.commons.security.saf`
+
+So you can do following:
+
+```java
+@GetMapping("/safProtectedResource")
+@PreAuthorize("hasSafResourceAccess('FACILITY', 'BPX.SERVER', 'UPDATE')")
```
Comment (dkelosky) [inline zowe-rest-api-sample-spring/docs/zos-security.md]: this is a really nice implementation!

**Comment by dkelosky** [comment]
I got: 
image

and:
image

And finished reading the document - this is a great feature!

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java (hunk 5 above)
Comment (plavjanik) [reply zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java]: This has been generated from the documentation - https://gist.github.com/plavjanik/a0f50e61edc2e2419a161e8d736e86de. Do you think that we should include the script into the source code?

**Comment by plavjanik** [comment]
> final comment - when do the new releases occur? after a merge to master?

Hi Dan, thank you a lot for catching all those problems! 

Releases occur when you create a new release on GitHub - https://github.com/zowe/sample-spring-boot-api-service/releases/new.

I tried to describe it at
https://github.com/zowe/sample-spring-boot-api-service/blob/master/docs/releasing.md.

Some projects publish a new snapshot version to JFrog OSS when there is a successful master build. We do not do that but might want to. It would be a way how we can provide an artifact to someone who needs a quick fix before we will do a new release.

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/service/AccessControlService.java (hunk 6 above)
Comment (plavjanik) [reply zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/service/AccessControlService.java]: yes, thank you

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/spring/security/SafMethodSecurityExpressionRoot.java (hunk 4 above)
Comment (plavjanik) [reply zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/spring/security/SafMethodSecurityExpressionRoot.java]: yes, thank you

File: zowe-rest-api-sample-spring/docs/zos-security.md (hunk 8 above)
Comment (plavjanik) [reply zowe-rest-api-sample-spring/docs/zos-security.md]: Thanks! I was considering to use existing Spring Security expression `hasAuthority()` but I was not sure how to put class, name, and level into one string without making it understandable so I end up creating a new expression.
https://docs.spring.io/autorepo/docs/spring-security/4.0.0.RC1/reference/html/el-access.html

File: zowe-rest-api-sample-spring/docs/zos-security.md (hunk 7 above)
Comment (plavjanik) [reply zowe-rest-api-sample-spring/docs/zos-security.md]: There are multiple typos and wrong word order (good in Czech but not in English). Sorry

File: zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java (hunk 5 above)
Comment (dkelosky) [reply zowe-rest-api-commons-spring/src/main/java/org/zowe/commons/zos/security/platform/PlatformErrno2.java]: ahh - that's a cool script!  that might be neat to include if its not much trouble or maybe just a link to your gist

**Comment by dkelosky** [comment]
no problems, just trivial, minor things.  awesome work - i'm anxious for others to try this out

**Pull request merged**